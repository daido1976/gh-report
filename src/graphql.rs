use graphql_client::GraphQLQuery;

#[allow(clippy::upper_case_acronyms)]
type URI = String;
type DateTime = chrono::DateTime<chrono::Utc>;

#[derive(GraphQLQuery)]
#[graphql(
    schema_path = "src/gql/schema.docs.graphql",
    query_path = "src/gql/me.gql",
    response_derives = "Debug"
)]
pub struct Me;

pub async fn exec(github_api_token: String) -> me::MeViewerContributionsCollection {
    println!("Hello, graphql!");
    let client = reqwest::Client::builder()
        .user_agent("graphql-rust/0.10.0")
        .default_headers(
            std::iter::once((
                reqwest::header::AUTHORIZATION,
                reqwest::header::HeaderValue::from_str(&format!("Bearer {}", github_api_token))
                    .unwrap(),
            ))
            .collect(),
        )
        .build()
        .unwrap();

    let variables = me::Variables {
        from_date_time: chrono::Utc::now() - chrono::Duration::days(10),
        to_date_time: chrono::Utc::now(),
    };

    let response_body = graphql_client::reqwest::post_graphql::<Me, _>(
        &client,
        "https://api.github.com/graphql",
        variables,
    )
    .await
    .unwrap();

    response_body.data.unwrap().viewer.contributions_collection
}
