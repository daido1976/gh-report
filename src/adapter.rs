use std::collections::HashMap;

use crate::graphql::{me::MeViewerContributionsCollection, NameWithOwner, URI};
use serde::Deserialize;

pub type MyContributions = HashMap<NameWithOwner, Vec<IssueOrPr>>;
#[derive(Debug, Deserialize)]
pub struct IssueOrPr {
    pub title: String,
    pub url: URI,
    pub state: String,
}

pub fn combine(contributions_collection: MeViewerContributionsCollection) -> MyContributions {
    let issue_contributions = contributions_collection.issue_contributions.edges.unwrap();
    let pr_contributions = contributions_collection
        .pull_request_contributions
        .edges
        .unwrap();

    let mut result: MyContributions = HashMap::new();
    for contribution in issue_contributions.into_iter().flatten() {
        let issue = contribution.node.unwrap().issue;
        let owner: NameWithOwner = issue.repository.name_with_owner;
        let issue = IssueOrPr {
            title: issue.title,
            url: issue.url,
            state: issue.state.to_string(),
        };
        result.entry(owner).or_insert_with(Vec::new).push(issue);
    }

    for contribution in pr_contributions.into_iter().flatten() {
        let pr = contribution.node.unwrap().pull_request;
        let owner: NameWithOwner = pr.repository.name_with_owner;
        let pr = IssueOrPr {
            title: pr.title,
            url: pr.url,
            state: pr.state.to_string(),
        };
        result.entry(owner).or_insert_with(Vec::new).push(pr);
    }
    result
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_combine() {
        let cc =
            serde_json::from_str(include_str!("fixtures/contributionsCollection.json")).unwrap();
        let actual = combine(cc);
        assert_debug_snapshot!(actual);
    }
}
