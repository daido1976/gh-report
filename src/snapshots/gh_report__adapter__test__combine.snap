---
source: src/adapter.rs
assertion_line: 56
expression: actual
---
{
    "daido1976/terakoya": [
        IssueOrPr {
            title: "From now to Vercel",
            url: "https://github.com/daido1976/terakoya/pull/52",
            state: "merged",
        },
    ],
    "rust-lang/rust": [
        IssueOrPr {
            title: "Cannot build on Fedora: wrong CPUTYPE?",
            url: "https://github.com/rust-lang/rust/issues/1218",
            state: "open",
        },
        IssueOrPr {
            title: "rustc: Fix position of diagnostic highlight lines",
            url: "https://github.com/rust-lang/rust/issues/1219",
            state: "closed",
        },
        IssueOrPr {
            title: "Add float support to #fmt.",
            url: "https://github.com/rust-lang/rust/pull/1168",
            state: "merged",
        },
        IssueOrPr {
            title: "Prohibit in-scope consts from use as variable names in binders, like nullary tags",
            url: "https://github.com/rust-lang/rust/pull/1193",
            state: "merged",
        },
    ],
}
